lib/scheduler/manual.go

```diff
@@ -119,51 +157,378 @@ func (mec ManualExecutionConfig) GetExecutionRequirements(es *lib.ExecutionSegme
+		}
+		//TODO: simulate spinner for the other case or cycle 0-100?
+		currentActiveVUs := atomic.LoadInt64(activeVUsCount)
+		currentMaxVUs := atomic.LoadInt64(maxVUs)
+		vusFmt := pb.GetFixedLengthIntFormat(currentMaxVUs)
+		return progress, fmt.Sprintf(
+			"currently "+vusFmt+" out of "+vusFmt+" active looping VUs, %s/%s", currentActiveVUs, currentMaxVUs,
+			pb.GetFixedLengthDuration(spent, duration), duration,
+		)
+	}
+	mex.progress.Modify(pb.WithProgress(progresFn))
+	go trackProgress(parentCtx, ctx, ctx, mex, progresFn)
+
+	currentlyPaused := false
+	waitVUs := func(from, to int64) {
+		for i := from; i < to; i++ {
+			vuHandles[i].wg.Wait()
+		}
+	}
+	handleConfigChange := func(oldControlConfig, newControlConfig ManualExecutionControlConfig) error {
+		oldActiveVUs := segment.Scale(oldControlConfig.VUs.Int64)
+		oldMaxVUs := segment.Scale(oldControlConfig.MaxVUs.Int64)
+		newActiveVUs := segment.Scale(newControlConfig.VUs.Int64)
+		newMaxVUs := segment.Scale(newControlConfig.MaxVUs.Int64)
+
+		mex.logger.WithFields(logrus.Fields{
+			"oldActiveVUs": oldActiveVUs, "oldMaxVUs": oldMaxVUs,
+			"newActiveVUs": newActiveVUs, "newMaxVUs": newMaxVUs,
+		}).Debug("Updating execution configuration...")
+
+		for i := oldMaxVUs; i < newMaxVUs; i++ {
+			vu, vuInitErr := mex.executorState.InitializeNewVU(ctx, mex.logger)
+			if vuInitErr != nil {
+				return vuInitErr
+			}
+			vuHandle := newManualVUHandle(
+				ctx, mex.executorState, activeVUsCount, vu, mex.logger.WithField("vuNum", i),
+			)
+			go vuHandle.runLoopsIfPossible(runIteration)
+			vuHandles = append(vuHandles, vuHandle)
+		}
+
+		if oldActiveVUs < newActiveVUs {
+			for i := oldActiveVUs; i < newActiveVUs; i++ {
+
+				if !currentlyPaused {
+					vuHandles[i].start()
+				}
+			}
+		} else {
+			for i := newActiveVUs; i < oldActiveVUs; i++ {
+				vuHandles[i].hardStop()
+			}
+			waitVUs(newActiveVUs, oldActiveVUs)
+		}
+
+		if oldMaxVUs > newMaxVUs {
+			for i := newMaxVUs; i < oldMaxVUs; i++ {
+				vuHandles[i].cancelVU()
+				if i < startMaxVUs {
+					// return the initial planned VUs to the common buffer
+					mex.executorState.ReturnVU(vuHandles[i].vu, false)
+				} else {
+					mex.executorState.ModInitializedVUsCount(-1)
+				}
+				vuHandles[i] = nil
+			}
+			vuHandles = vuHandles[:newMaxVUs]
+		}
+
+		atomic.StoreInt64(maxVUs, newMaxVUs)
+		return nil
+	}
+
+	err = handleConfigChange(ManualExecutionControlConfig{MaxVUs: mex.startConfig.MaxVUs}, currentControlConfig)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = handleConfigChange(currentControlConfig, ManualExecutionControlConfig{})
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case updateConfigEvent := <-mex.newControlConfigs:
+			err := handleConfigChange(currentControlConfig, updateConfigEvent.newConfig)
+			if err != nil {
+				updateConfigEvent.err <- err
+				return err
+			}
+			currentControlConfig = updateConfigEvent.newConfig
```

Comment:
At this point, `mex.currentControlConfig` is just saved so that the `GetCurrentConfig()` method can function correctly, so

>  just configLock/Unlock around every time we get mex.currentControlConfig and use it instead

this would just add unnecessary synchronization in the code